tests/unit/test_hsm.cpp

```diff
@@ -74,32 +83,45 @@ void HSMTest::SetUp()
 
 void HSMTest::TearDown() { openssl::OpenSSLLibMockManager::destroy(); }
 
-TEST_F(HSMTest, testHSMEngine)
+std::unique_ptr<HsmEngine> HSMTest::initialiseEngine()
 {
     std::string engineID("engine_id");
     std::string modulePath("/test_path.so");
     std::string pin("1234");
+    auto engine = ::testutils::someEnginePtr();
 
-    EXPECT_CALL(_mock(), SSL_ENGINE_by_id(StrEq(engineID.c_str())))
-            .WillOnce(Return(::testutils::someEnginePtr()));
+    EXPECT_CALL(_mock(), SSL_ENGINE_by_id(StrEq(engineID.c_str()))).WillOnce(Return(engine));
 
-    EXPECT_CALL(_mock(),
-                SSL_ENGINE_ctrl_cmd_string(::testutils::someEnginePtr(),
-                                           StrEq("MODULE_PATH"),
-                                           StrEq(modulePath.c_str()),
-                                           0 /*non-optional*/))
+    EXPECT_CALL(
+            _mock(),
+            SSL_ENGINE_ctrl_cmd_string(
+                    engine, StrEq("MODULE_PATH"), StrEq(modulePath.c_str()), 0 /*non-optional*/))
             .WillOnce(Return(1));
 
     EXPECT_CALL(_mock(),
-                SSL_ENGINE_ctrl_cmd_string(::testutils::someEnginePtr(),
-                                           StrEq("PIN"),
-                                           StrEq(pin.c_str()),
-                                           0 /*non-optional*/))
+                SSL_ENGINE_ctrl_cmd_string(
+                        engine, StrEq("PIN"), StrEq(pin.c_str()), 0 /*non-optional*/))
             .WillOnce(Return(1));
 
-    EXPECT_CALL(_mock(), SSL_ENGINE_init(::testutils::someEnginePtr())).WillOnce(Return(1));
+    EXPECT_CALL(_mock(), SSL_ENGINE_init(engine)).WillOnce(Return(1));
+    EXPECT_CALL(_mock(), SSL_ENGINE_finish(engine)).WillOnce(Return(1));
 
-    EXPECT_CALL(_mock(), SSL_ENGINE_finish(::testutils::someEnginePtr())).WillOnce(Return(1));
+    return std::make_unique<HsmEngine>(engineID, modulePath, pin);
+}
 
-    EXPECT_NO_THROW((HsmEngine(engineID, modulePath, pin)));
+TEST_F(HSMTest, testHSMKeygen)
+{
+    ECCSpec eccSpec;
+    int curve = int(mococrw::openssl::ellipticCurveNid::PRIME_256v1);
+    auto engine = ::testutils::someEnginePtr();
+    auto pkey = ::testutils::somePkeyPtr();
+    auto hsm = initialiseEngine();
+    EXPECT_CALL(_mock(), SSL_EC_curve_nid2nist(curve)).WillOnce(Return("P-256"));
+    EXPECT_CALL(_mock(),
+                SSL_ENGINE_ctrl_cmd(engine, StrEq("KEYGEN"), 0 /*non-optional*/, _, nullptr, 1))
```

Comment:
If we would want to be more thorough, we would also match the struct that is passed to the engine command. But I guess for now its ok to have it like this. In particular, given that engines are deprecated and to be replaced by providers anyways.